DESIGN.md

```diff
@@ -0,0 +1,211 @@
+                "jobStatus": "stopped",
+                "owner": "user@example.com",
+                "startTime": "2020-08-18T12:00:00Z", 
+                "stopTime": "2020-08-18T12:00:00Z" ,
+                "exitCode": 0
+                "stderrSize": 123,
+                "stderrBytes": "bytes",
+                "stdoutSize": 321,
+                "stdoutBytes": "more bytes"
+            }
+            ```
+            - Error response: 
+            ```json
+            {"status": 404, "error": "no job found with that ID"}
+            ```
+        - **GET /list** gets a list of all of the user's jobs and their status. Will not return jobs 
+        the user does not own.
+            - Success response:
+            ```json
+            {
+                "status": 200, 
+                "jobs": [
+                    {
+                        "jobID": 123,
+                        "job": "ls -lh /var/log",
+                        "owner": "user@example.com",
+                        "jobStatus": "stopped",
+                        "startTime": "2020-08-18T12:00:00Z", 
+                        "stopTime": "2020-08-18T12:00:00Z" ,
+                    },
+                    {
+                        "jobID": 124,
+                        "job": "ticker -interval=1m",
+                        "owner": "user@example.com",
+                        "jobStatus": "running",
+                        "startTime": "2020-08-18T12:00:00Z", 
+                    }
+                ]
+            }
+            ```
+            - Error response: 
+            ```json
+            {"status": 500, "error": "some failure"}
+            ```
+    - *cli*: the CLI tool for managing jobs. Hard-coded credentials and server address.
+        - Commands will include `start`, `stop`, `status`, `output`, and `list` just like the HTTP API.
+    - *worker*: The actual Linux service to manage jobs. Using a service is convenient because it provides 
+    automatic restarts if the service crashes due to somebody running something that consumes enormous RAM or 
+    something.
+        - Exposes the methods for Start, Stop, Status, Output, List via RPC
+- *Makefile*: will build one or all of the 3 executables
+
+
+### Worker
+The job worker will make use of `net/rpc` for communicating with the HTTP server. The package is in feature lock, but 
+provides an easy interface for IPC and development is still current. There have been new commits to the package that
+will be included in the Go 1.17 release. Using Unix domain sockets will allow pretty low latency and is supported by 
+net/rpc. Better performance could be possible, but won't be necessary for this example. 
+
+The `os/exec` package allows capturing stdout/stderr directly and store it in the job. Not memory efficient to keep it 
+all in RAM, but quick to implement. Storing stdout and stderr will be done separately to allow the user to get get 
+them as two separate fields when retreiving output.
```

Comment:
On the client side how will the output be displayed if receiving two separate fields?

Reply:
The API will return them as two separate fields `stderrBytes` and `stdoutBytes` and the CLI will perhaps use something like this in the output:
```
stdout: 
    line of data
    line of data
stderr:
    error 1
    error 2
```

I had not considered it until now, but I was returning `stdoutBytes` rather than `stdoutString` so that the client could get back non-string data (perhaps an image). That won't work with JSON directly so it would have to be encoded, probably to base64. In that case it may be better to just call the fields `stdout` and `stderr` and document that the data will be encoded as base64.

Reply:
You can still return bytes if you just have your /output/jobID endpoint return pure bytes instead of json. Then no encoding will be needed. All that other information in the /output endpoint is available to the client as a part of /status call.